alf/algorithms/off_policy_algorithm.py

```diff
@@ -206,3 +219,266 @@ def preprocess_experience(self, experience: Experience):
+                    input_params_spec={"loc": spec},
+                    sample_spec=spec)
+            return spec
+
+        self._action_distribution_spec = tf.nest.map_structure(
+            _to_distribution_spec, self.action_distribution_spec)
+        self._action_dist_param_spec = tf.nest.map_structure(
+            lambda spec: spec.input_params_spec,
+            self._action_distribution_spec)
+
+        self._experience_spec = Experience(
+            step_type=self._time_step_spec.step_type,
+            reward=self._time_step_spec.reward,
+            discount=self._time_step_spec.discount,
+            observation=self._time_step_spec.observation,
+            prev_action=self._action_spec,
+            action=self._action_spec,
+            info=info_spec,
+            action_distribution=self._action_dist_param_spec,
+            state=self.train_state_spec if self._use_rollout_state else ())
+
+        if exp_replayer == "one_time":
+            self._exp_replayer = OnetimeExperienceReplayer()
+        elif exp_replayer == "uniform":
+            self._exp_replayer = SyncUniformExperienceReplayer(
+                self._experience_spec, env_batch_size)
+        else:
+            raise ValueError("invalid experience replayer name")
+        self.add_experience_observer(self._exp_replayer.observe)
+
+        action_dist_params = common.zero_tensor_from_nested_spec(
+            self._experience_spec.action_distribution, env_batch_size)
+        action_dist = nested_distributions_from_specs(
+            self._action_distribution_spec, action_dist_params)
+
+        exp = Experience(
+            step_type=time_step.step_type,
+            reward=time_step.reward,
+            discount=time_step.discount,
+            observation=time_step.observation,
+            prev_action=time_step.prev_action,
+            action=time_step.prev_action,
+            info=policy_step.info,
+            action_distribution=action_dist,
+            state=initial_state if self._use_rollout_state else ())
+
+        processed_exp = self.preprocess_experience(exp)
+        self._processed_experience_spec = self._experience_spec._replace(
+            info=extract_spec(processed_exp.info))
+
+        policy_step = common.algorithm_step(
+            algorithm_step_func=self.train_step,
+            time_step=exp,
+            state=initial_state)
+        info_spec = extract_spec(policy_step.info)
+        self._training_info_spec = make_training_info(
+            action=self._action_spec,
+            action_distribution=self._action_dist_param_spec,
+            step_type=self._time_step_spec.step_type,
+            reward=self._time_step_spec.reward,
+            discount=self._time_step_spec.discount,
+            info=info_spec,
+            collect_info=self._processed_experience_spec.info,
+            collect_action_distribution=self._action_dist_param_spec)
+
+    @tf.function
+    def train(self,
```

Comment:
To make things more flexible, I feel obtaining data from exp_player should also be a part of job of train()